[DPA-1525] solana ccip transfer ownership to timelock

ecPablo opened the pull request:
https://smartcontract-it.atlassian.net/browse/DPA-1525

Adds the changeset to transfer ownership of CCIP solana programs. Should be merged after 

* https://github.com/smartcontractkit/chainlink/pull/16159 as FeeQuoter branch was merged here in order to get access to new bindings
* https://github.com/smartcontractkit/chainlink/pull/16163 to have the deployment changesets avaialble

TODO:
- [x] Add Offramp Transfer
- [x] Add Timelock Execution to tests

## AI Summary
This pull request includes several changes to the `deployment/ccip/changeset` package, focusing on improvements and updates to the Solana chain contracts and related tests. The most important changes include updating import statements, modifying function parameters, and adding new functions for initializing and deploying contracts. Additionally, there are updates to the test files to reflect these changes.

### Changes to Solana Chain Contracts:

* **Import Statements:**
  * Reordered and grouped import statements in `cs_chain_contracts.go` and `cs_deploy_chain.go` for better organization. [[1]](diffhunk://#diff-674af96625f8adc0939a49250b0e58ddb26c32312f344213353090fc809f3429L21-R33) [[2]](diffhunk://#diff-b7f8a498937d8fcd6b2d3850cfe63f19fec5932061406345cf5f22e9c8433b11R20-L24)

* **Function Modifications:**
  * Updated `validateRemoteChain` function to use `chainState.Timelock.ProgramID.String()` instead of `chainState.Timelock`.
  * Modified `DefaultFeeQuoterDestChainConfig` function to accept an optional `destChainSelector` parameter and added logic to handle Solana chain selectors. [[1]](diffhunk://#diff-674af96625f8adc0939a49250b0e58ddb26c32312f344213353090fc809f3429L1618-R1633) [[2]](diffhunk://#diff-674af96625f8adc0939a49250b0e58ddb26c32312f344213353090fc809f3429L1644-R1651)
  * Added `feeQuoterAddress` parameter to `initializeRouter` function and updated its logic accordingly. [[1]](diffhunk://#diff-b7f8a498937d8fcd6b2d3850cfe63f19fec5932061406345cf5f22e9c8433b11L559-R557) [[2]](diffhunk://#diff-b7f8a498937d8fcd6b2d3850cfe63f19fec5932061406345cf5f22e9c8433b11L572-R572) [[3]](diffhunk://#diff-b7f8a498937d8fcd6b2d3850cfe63f19fec5932061406345cf5f22e9c8433b11R595-R624)

* **New Functions:**
  * Introduced `initializeFeeQuoter` function to initialize the fee quoter for Solana chains.

### Changes to Tests:

* **Import Statements:**
  * Updated import statements in `cs_chain_contracts_test.go` to include new dependencies and remove unused ones.

* **Function Updates:**
  * Updated `TestAddRemoteChain` and `TestAddTokenPool` functions to align with the changes in the main codebase, including the use of new structures and configuration parameters. [[1]](diffhunk://#diff-63c14720b1d34153c81bdcd83289635be7603869be209d97a1c988ae2fea5854L36-R41) [[2]](diffhunk://#diff-63c14720b1d34153c81bdcd83289635be7603869be209d97a1c988ae2fea5854L57-R60) [[3]](diffhunk://#diff-63c14720b1d34153c81bdcd83289635be7603869be209d97a1c988ae2fea5854L77-R79) [[4]](diffhunk://#diff-63c14720b1d34153c81bdcd83289635be7603869be209d97a1c988ae2fea5854L168-R187)
  * Added context initialization in `TestAddTokenPool` and `TestBilling` functions. [[1]](diffhunk://#diff-63c14720b1d34153c81bdcd83289635be7603869be209d97a1c988ae2fea5854L103-R103) [[2]](diffhunk://#diff-63c14720b1d34153c81bdcd83289635be7603869be209d97a1c988ae2fea5854L219-R223)

tt-cll commented on deployment/ccip/changeset/solana/ownership_transfer_helpers.go:
```diff
@@ -0,0 +1,266 @@
+	configPDA solana.PublicKey, // e.g. for routerConfigPDA or a token-pool config
+	deployer solana.PublicKey, // the “from” authority
+	solChain deployment.SolChain, // used for solChain.Confirm
+	label string, // e.g. "Router" or "TokenPool"
```
might make it easier to debug if we make this a deployment.ContractType and then just cast it to a string when passing to the mcms lib

tt-cll replied:
i.e. https://github.com/smartcontractkit/chainlink/blob/f84bbe0a67f410578b14ea65ea2f488e045f14a7/deployment/ccip/changeset/solana_state.go#L18

ecPablo replied:
good call! done here https://github.com/smartcontractkit/chainlink/pull/16246/commits/7e8c23a660693ae111b66a89b3fbbc12020c2556

tt-cll commented on deployment/ccip/changeset/solana/ownership_transfer_helpers.go:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	// 3. Build the “accept ownership” instruction
+	ixAccept, err := buildAccept(configPDA, proposedOwner)
```
What happens if the transfer succeeds and then we fail somewhere in the accept? We can just retry with no issues?

ecPablo replied:
Yes the transfer just sets a `proposed_owner` field on the program's config. If accept fails you can alway call it again, or call transferOwnership again to re-set `proposed_owner` to something else.

tt-cll commented on deployment/ccip/changeset/solana/ownership_transfer_helpers.go:
```diff
@@ -0,0 +1,266 @@
+	mcmsTx, err := mcmsSolana.NewTransaction(
+		programID.String(),
+		acceptData,
+		big.NewInt(0),       // e.g. value
```
is this always going to be 0? Do we need to put this in a param?

ecPablo replied:
Yeah mot of the time it will be unless we are doing a sol transfer, but for the transfer ownership case we can leave it as 0.

tt-cll commented on deployment/ccip/changeset/solana/ownership_transfer_helpers.go:
```diff
@@ -0,0 +1,266 @@
+		).ValidateAndBuild()
+	}
+	buildAccept := func(config, newOwnerAuthority solana.PublicKey) (solana.Instruction, error) {
+		ccip_router.SetProgramID(routerProgramID)
```
is this working for you? I'm almost positive we hit errors if we set the program id multiple times for the same program

graham-chainlink replied:
I believe it does work, code just ignores it 

https://github.com/gagliardetto/solana-go/blob/main/registry.go#L83

ecPablo replied:
Yeah for me it's not giving issues. But it would be a good idea if we could all agree where to place these `SetProgramID` on the changesets. Right now its kind of arbitrary

tt-cll commented on deployment/ccip/changeset/solana/ownership_transfer_helpers.go:
```diff
@@ -0,0 +1,266 @@
+
+	result = append(result, tx)
+	return result, nil
+}
```
remind me: we said we need clarity on if MCMS will also support transfer and accept ownership for token pools? Because we do have those instructions available there too

ecPablo replied:
Yeah, we just need to confirm if we do need to transfer ownership of those to the timelock or not. If it's the case then I can create a couple more calls to transfer those in a followup.

tt-cll commented on deployment/ccip/changeset/solana_state.go:
```diff
@@ -154,7 +152,7 @@ func LoadChainStateSolana(chain deployment.SolChain, addresses map[string]deploy
 			state.OffRampStatePDA = offRampStatePDA
 		default:
-			return state, fmt.Errorf("unknown contract %s", tvStr)
+			continue
```
We should at least log a warning or something

ecPablo replied:
added here https://github.com/smartcontractkit/chainlink/pull/16246/commits/7e8c23a660693ae111b66a89b3fbbc12020c2556

tt-cll commented on deployment/common/changeset/internal/solana/access_controller.go:
```diff
@@ -82,13 +83,12 @@ func initAccessController(
 
-	address := state.EncodeAddressWithAccount(programID, account.PublicKey())
-	err = addressBook.Save(chain.Selector, address, typeAndVersion)
+	err = addressBook.Save(chain.Selector, account.PublicKey().String(), typeAndVersion)
```
I think we need a corresponding entry in the state to read this back, right?

ecPablo replied:
Yeah we do have it here: https://github.com/smartcontractkit/chainlink/blob/d080e78db4f663981fdd4b7a28d5c358c5a023b5/deployment/common/changeset/state/solana.go#L22 it's just separated as it's part of the `common` package since this one is product agnostic.

tt-cll commented on deployment/environment.go:
```diff
@@ -47,6 +47,10 @@ type OffchainClient interface {
 	nodev1.NodeServiceClient
 	csav1.CSAServiceClient
 }
+type MultiFamilyChains struct {
```
do we still need this?

ecPablo replied:
nope, removed thanks! https://github.com/smartcontractkit/chainlink/pull/16246/commits/7e8c23a660693ae111b66a89b3fbbc12020c2556

gustavogama-cll commented on deployment/ccip/changeset/testhelpers/test_helpers.go:
```diff
@@ -1642,7 +1643,7 @@ func DeployLinkTokenTest(t *testing.T, solChains int) {
 	addrs, err := e.ExistingAddresses.AddressesForChain(chain1)
 	require.NoError(t, err)
-	state, err := commoncs.MaybeLoadLinkTokenChainState(e.Chains[chain1], addrs)
+	state, err := commonState.MaybeLoadLinkTokenChainState(e.Chains[chain1], addrs)
```
I was hoping we would be able to avoid migrating the old changesets and helpers to the new state package. Alas... fingers crossed.

ecPablo replied:
I think this might be the only one we need to migrate as it's also used by the solana changesets but we can leave the other ones as it is

gustavogama-cll commented on deployment/common/changeset/internal/solana/access_controller.go:
```diff
@@ -82,13 +83,12 @@ func initAccessController(
 
-	address := state.EncodeAddressWithAccount(programID, account.PublicKey())
-	err = addressBook.Save(chain.Selector, address, typeAndVersion)
+	err = addressBook.Save(chain.Selector, account.PublicKey().String(), typeAndVersion)
```
we're deviating from the pattern we adopted for the mcm contracts, where we store both the program id as well as the seed in the address book entry.

I think it still works if you save the program id as a separate entry in the address book, but we need to be careful as we're being inconsistent and using a different standard for each program.

ecPablo replied:
I believe for the access controller the case is a bit different because here it's actually separate public keys and not [32]byte seeds. Since each access controller role is created with a create account + an initialize instruction. For the mcms and timelock programs we are following the same pattern

gustavogama-cll replied:
yeah, but the seeds are also `[32]byte`, so I cheated and used the same setter function. That's not what I'm concerned about though -- we can fix the setters and make them fit the data, if needed.

What I was trying to hint at is that, for mcm and timelock, the address book entries (say `ProposerMcm`) contain both the program id as well as the additional data you need to identify that particular instance (the seed, in mcm's case). But for one of the access controllers, we only store the account in the address book:

```
AddressBook {
  ManyChainMultisigProgram:        "mcmProgramID",
  ProposerManyChainMultisig:       "mcmProgramID.proposerSeed",
  CancellerManyChainMultiSig:      "mcmProgramID.cancellerSeed",
  ...
  RBACTimelockProgram:             "timelockProgramID",
  RBACTimelock:                    "timelockProgramID.timelockSeed",
  AccessControllerProgram:         "accessControllerProgramID",
  ProposerAccessControllerAccount: "proposerAccount",  // as opposed to "accessControllerProgramID.proposerAccount"
  ExecutorAccessControllerAccount: "executorAccount",  // as opposed to "accessControllerProgramID.executorAccount"
  ...
}
```

We can work with it, no information is *missing*. But I wanted to point out it's inconsistent. And that I implemented it as such originally to avoid the inconsistency. :-)

ecPablo replied:
Ahh I see now, but still I think conceptually these are still separate accounts right? In the case of the timelock and mcms programs, the seeds are used to determine the instance of the state and used for deriving all the pdas (which are owned by the mcms/timelock). But in the case of the access controller, these are separate accounts and not used to derive any PDAs so in my head, it was kind of breaking the patterns `programID.seed` because in this case they are not seeds but solana accounts. But I do see it can make sense to keep the pattern as they are related in the sense that they are owned by the access controller program.

gustavogama-cll commented on deployment/common/changeset/mcmsnew/deploy_mcms_with_timelock.go:
something is off. Why is this file here? It's outdated and we already have `deployment/common/changeset/deploy_mcms_with_timelock.go`.

gustavogama-cll replied:
I think the whole `mcmsnew` folder is duplicated with an old version.

gustavogama-cll replied:
And I'm wondering if the updates related to the `state` package were also inherited from an old version of my PR...

ecPablo replied:
Yeah I think there's something off with the merge conclicts resolution. I'm going to recheck your branch to make sure everything is updated.

ecPablo replied:
this should be better now https://github.com/smartcontractkit/chainlink/pull/16246/commits/042f3127c765983c89328bc9a8bc6a09967e3a87

gustavogama-cll commented on deployment/common/changeset/state.go:
yes, I'm now 99% sure the state changes were from an old version of my PR. This file shouldn't be deleted, imo.

gustavogama-cll commented on deployment/common/changeset/state/solana.go:
```diff
@@ -235,35 +232,31 @@ func MaybeLoadMCMSWithTimelockChainStateSolana(chain deployment.SolChain, addres
+			account, err := solana.PublicKeyFromBase58(address)
 			if err != nil {
-				return nil, fmt.Errorf("unable to decode canceller access controlle address (%s): %w", address, err)
+				return nil, fmt.Errorf("unable to decode executor access controller address (%s): %w", address, err)
```
canceller

ecPablo replied:
thanks ! https://github.com/smartcontractkit/chainlink/pull/16246/commits/042f3127c765983c89328bc9a8bc6a09967e3a87

gustavogama-cll commented on deployment/common/changeset/test_helpers.go:
```diff
@@ -133,3 +138,103 @@ func ApplyChangesets(t *testing.T, e deployment.Environment, timelockContractsPe
 }
+
+// ApplyChangesetsV2 applies the changeset applications to the environment and returns the updated environment.
+func ApplyChangesetsV2(t *testing.T, e deployment.Environment, changesetApplications []ConfiguredChangeSet) (deployment.Environment, error) {
```
why do we need an ApplyChangesetsV2? Doesn't the original support the new mcms lib?

ecPablo replied:
The original one required we provide the set of timelock addresses per chain, but on the new one we can simplify that as we can extract the timelocks from the proposal

```go
func ApplyChangesets(t *testing.T, e deployment.Environment, timelockContractsPerChain map[uint64]*proposalutils.TimelockExecutionContracts, changesetApplications []ConfiguredChangeSet) (deployment.Environment, error)
```

to this
```go
func ApplyChangesetsV2(t *testing.T, e deployment.Environment, changesetApplications []ConfiguredChangeSet) (deployment.Environment, error)
```

gustavogama-cll commented on deployment/environment/memory/chain.go:
```diff
@@ -118,6 +119,19 @@ func generateSolanaKeypair(t testing.TB) (solana.PrivateKey, string, error) {
 
 	return privateKey, keypairPath, nil
 }
+func FundTestAccounts(t *testing.T, keys []solana.PublicKey, url string) {
```
whitespace before func

ecPablo replied:
fixed! https://github.com/smartcontractkit/chainlink/pull/16246/commits/042f3127c765983c89328bc9a8bc6a09967e3a87

gustavogama-cll commented on deployment/environment/memory/chain.go:
```diff
@@ -118,6 +119,19 @@ func generateSolanaKeypair(t testing.TB) (solana.PrivateKey, string, error) {
+		_, err := exec.Command("solana", "airdrop", "100",
+			account,
+			"--url", url,
+			"--commitment", "finalized",
```
this will slow down tests significantly; using `finalized` means we need to wait for ~32 slots, which is ~12 seconds.

ecPablo replied:
This is a mistake, I added it as of debugging the tests as I was having some confirmation issues with the state, but I just removed it. Thanks!

gustavogama-cll commented on deployment/go.mod:
```diff
@@ -39,7 +39,7 @@ require (
 	github.com/smartcontractkit/chainlink-testing-framework/lib v1.50.22
 	github.com/smartcontractkit/libocr v0.0.0-20241223215956-e5b78d8e3919
-	github.com/smartcontractkit/mcms v0.10.0
+	github.com/smartcontractkit/mcms v0.9.0
```
is this correct? we're downgrading?

ecPablo replied:
nope, this should stay on the same version thanks!

ecPablo replied:
https://github.com/smartcontractkit/chainlink/pull/16246/commits/042f3127c765983c89328bc9a8bc6a09967e3a87

gustavogama-cll commented on deployment/common/proposalutils/mcms_helpers.go:
```diff
@@ -10,12 +10,86 @@ import (
+}
+
+// BuildInspectorsFromEnv generates a map of Inspectors for each chain in the environment.
+func BuildInspectorsFromEnv(e deployment.Environment) (map[mcmstypes.ChainSelector]sdk.Inspector, error) {
```
Maybe we could use [these](https://github.com/smartcontractkit/chainlink/pull/16313/files#diff-21c53969f9bc869d7b6ec247331547656e67077caf443c36c999b98088bebec2R283-R362) instead? I think they're a bit more flexible and composible.

And maybe we can apply the same pattern for the `BuildConvertersFromEnv` helper above, and have a `McmsConverters` and `McmsConverterForChain`?

gustavogama-cll replied:
btw, I have an open PR ready to merge the helpers I mentioned:
https://github.com/smartcontractkit/chainlink/pull/16403

ecPablo replied:
I just removed the helpers I had in favor of your PR. They were not used anymore so just had to remove it from the `mcms_helpers.go`

gustavogama-cll commented on deployment/common/proposalutils/mcms_test_helpers.go:
```diff
@@ -181,57 +182,123 @@ func SignMCMSProposal(t *testing.T, env deployment.Environment, proposal *mcmsli
-// ExecuteMCMSProposalV2 - Executes an MCMS proposal on a chain. For timelock proposal, use ExecuteMCMSTimelockProposalV2 instead.
 func ExecuteMCMSProposalV2(t *testing.T, env deployment.Environment, proposal *mcmslib.Proposal, sel uint64) {
-	t.Log("Executing proposal on chain", sel)
+	t.Logf("[ExecuteMCMSProposalV2] Executing proposal on chain %d...", sel)
```
this looks outdated...

ecPablo replied:
I did add this log to the V2 version, it was helpful to debug the in memory tests and I'd say we could keep them. But let me know if you're against it

gustavogama-cll replied:
I wasn't referring to the logs. I was referring to the function itself. See what the [current version](https://github.com/smartcontractkit/chainlink/blob/develop/deployment/common/proposalutils/mcms_test_helpers.go#L175-L240) looks like.

gustavogama-cll replied:
[ExecuteMCMSTimelockProposalV2](https://github.com/smartcontractkit/chainlink/blob/develop/deployment/common/proposalutils/mcms_test_helpers.go#L175-L240) might be outdated as well.

gustavogama-cll commented on deployment/common/changeset/test_helpers.go:
```diff
@@ -114,3 +119,98 @@ func ApplyChangesets(t *testing.T, e deployment.Environment, timelockContractsPe
+	}
+	return currentEnv, nil
+}
+func DeployLinkTokenTest(t *testing.T, solChains int) {
```
space before new function

ecPablo replied:
fixed https://github.com/smartcontractkit/chainlink/pull/16246/commits/db3ba9ecb9cb9f48aa91494f2fbc68c3785b7e31

yashnevatia commented on deployment/environment/memory/chain.go:
```diff
@@ -119,6 +120,19 @@ func generateSolanaKeypair(t testing.TB) (solana.PrivateKey, string, error) {
 	return privateKey, keypairPath, nil
 }
 
+func FundTestAccounts(t *testing.T, keys []solana.PublicKey, url string) {
```
can we rename to FundTestAccountsWithSol ?

ecPablo replied:
Done! https://github.com/smartcontractkit/chainlink/pull/16246/commits/7e8dfdbdf9381d0a8f54bcc4d2350894e864b74f

yashnevatia commented on deployment/ccip/changeset/solana/transfer_ccip_to_mcms_with_timelock_test.go:
```diff
@@ -0,0 +1,384 @@
+	nodes, err := deployment.NodeInfo(e.NodeIDs, e.Offchain)
+	require.NoError(t, err)
+	// Fund account for fees
+	memory.FundTestAccounts(t, []solana.PublicKey{solChain.DeployerKey.PublicKey()}, solChain.URL)
```
i think if we have the chain.Client already available, we should consider [this](https://github.com/smartcontractkit/chainlink-ccip/blob/3fe6ba52c5d84dbc48c6ca777c195dfb53deef7f/chains/solana/contracts/tests/testutils/anchor.go#L25) ?

ecPablo replied:
Yes I've replaced it here too https://github.com/smartcontractkit/chainlink/pull/16246/commits/7e8dfdbdf9381d0a8f54bcc4d2350894e864b74f

yashnevatia commented on deployment/common/changeset/state/pda.go:
```diff
@@ -0,0 +1,75 @@
+package state
```
nit: can we move this [here](https://github.com/smartcontractkit/chainlink-ccip/blob/main/chains/solana/utils/state/pda.go) so that we have all of this in one place ?

ecPablo replied:
I think those helper already exist in chainlink-ccip [here](https://github.com/smartcontractkit/chainlink-ccip/blob/cd1c128d597d54a96a58fedb6220fb24a4698cb9/chains/solana/utils/mcms/mcm.go#L16) however the problem is that they are using a hardcoded programID. We did not want to mess with the contract tests so we opted for having them here, but we can move it if you find it valuable in a followup?

yashnevatia replied:
yes lets move it as a follow up.
and you can raise a PR in chainlink-ccip so that it uses the function as well like terry did [here](https://github.com/smartcontractkit/chainlink-ccip/pull/497)

yashnevatia commented on deployment/ccip/changeset/cs_chain_contracts.go:
```diff
@@ -1341,13 +1342,24 @@ func (c SetOCR3OffRampConfig) validateRemoteChain(e *deployment.Environment, sta
+		if err != nil {
+			return fmt.Errorf("error loading MCMS state for chain %d: %w", chainSelector, err)
+		}
 		// TODO: introduce interface when MCMS is ready
```
nit: should we remove the TODO ?

ecPablo replied:
removed it here! also moved the todo to the implementation since we have not implemented the validate ownership function yet https://github.com/smartcontractkit/chainlink/pull/16246/commits/7e8dfdbdf9381d0a8f54bcc4d2350894e864b74f

yashnevatia commented on deployment/ccip/changeset/solana/transfer_ccip_to_mcms_with_timelock_test.go:
```diff
@@ -0,0 +1,384 @@
+	ctx := context.Background()
+
+	// (A) Check Router ownership -  we need to add retries as the ownership transfer commitment is confirmed and not finalized.
+	require.Eventually(t, func() bool {
```
i think can use [this](https://github.com/smartcontractkit/chainlink/blob/1c20a04df3ea25a5a90ad71b3f70b5f7eda47b81/deployment/ccip/changeset/solana/cs_chain_contracts_test.go#L91) instead (it does the unmarshalling) ?

yashnevatia replied:
eg:
```golang
tokenAdminRegistryPDA, _, _ := solState.FindTokenAdminRegistryPDA(tokenAddress, state.SolChains[solChain].Router)
	var tokenAdminRegistryAccount solRouter.TokenAdminRegistry
	err = e.SolChains[solChain].GetAccountDataBorshInto(ctx, tokenAdminRegistryPDA, &tokenAdminRegistryAccount)
	require.NoError(t, err)
	require.Equal(t, solana.PublicKey{}, tokenAdminRegistryAccount.Administrator)
	// pending administrator should be the proposed admin key
	require.Equal(t, tokenAdminRegistryAdminPrivKey.PublicKey(), tokenAdminRegistryAccount.PendingAdministrator)
```

ecPablo replied:
Yes thanks! this simplifies things a lot https://github.com/smartcontractkit/chainlink/pull/16246/commits/7e8dfdbdf9381d0a8f54bcc4d2350894e864b74f

yashnevatia commented on deployment/ccip/changeset/solana/transfer_ccip_to_mcms_with_timelock_test.go:
```diff
@@ -0,0 +1,384 @@
+		})
+	}
+	testhelpers.SavePreloadedSolAddresses(t, e, solChainSelectors[0])
+	e, err = commonchangeset.ApplyChangesets(t, e, nil, []commonchangeset.ConfiguredChangeSet{
```
can use `tenv, _ := testhelpers.NewMemoryEnvironment(t, testhelpers.WithSolChains(1))` this instead ?

ecPablo replied:
We do use it in the `prepareEnvironmentForOwnershipTransfer()` here we just receive the same updated env after the changeset application.

yashnevatia replied:
isnt the function i commented on `prepareEnvironmentForOwnershipTransfer` ?

yashnevatia commented on deployment/ccip/changeset/solana/transfer_ccip_to_mcms_with_timelock_test.go:
```diff
@@ -0,0 +1,384 @@
+		routerAcc, err := solClient.GetAccountInfo(ctx, routerConfigPDA)
+		require.NoError(t, err, "failed to fetch Router config account info")
+		programData := ccip_router.Config{}
+		ccip_router.SetProgramID(state.SolChains[solChain1].Router)
```
why do we need to setProgramId here ? i dont see ccip_router being used anywhere ?

ecPablo replied:
I think we did not need it, removed it as we are using now the helper you suggested https://github.com/smartcontractkit/chainlink/pull/16246/commits/7e8dfdbdf9381d0a8f54bcc4d2350894e864b74f

yashnevatia commented on deployment/ccip/changeset/solana/transfer_ccip_to_mcms_with_timelock.go:
```diff
@@ -0,0 +1,218 @@
+			return fmt.Errorf("mcms proposer not present on the chain %w", err)
+		}
+		// Make sure addresses are correctly parsed. Format is: "programID.PDASeed"
+		_, _, err = mcmsSolana.ParseContractAddress(timelockID)
```
have not gone in-depth here, but can we use the address book labels ?

yashnevatia replied:
```golang
case TokenPoolLookupTable:
			lookupTablePubKey := solana.MustPublicKeyFromBase58(address)
			// Labels should only have one entry
			for tokenPubKeyStr := range tvStr.Labels {
				tokenPubKey := solana.MustPublicKeyFromBase58(tokenPubKeyStr)
				state.TokenPoolLookupTable[tokenPubKey] = lookupTablePubKey
			}
```

ecPablo replied:
We decided to keep it as a string with Gustavo as it's also the same format that the mcms go library accepts, just to keep things consistent between the repos. But we can think if we can use labels for this in a followup, the address book will be refactored so we can do it after it is a bit more stable.

gustavogama-cll replied:
I'm not too fond of using the labels for this. At least not with the current implementation, which I find too constraining. Yash's example even says so:
```go
// Labels should only have one entry
```

gustavogama-cll commented on deployment/ccip/changeset/cs_chain_contracts.go:
```diff
@@ -18,6 +18,7 @@ import (
 
 	"github.com/smartcontractkit/chainlink-common/pkg/logger"
 
+	commonState "github.com/smartcontractkit/chainlink/deployment/common/changeset/state"
```
pablo, the state modifications are still part of the PR

ecPablo committed:
feat(deployment): solana ccip transfer ownership to timelock

Adds the changeset to transfer ownership of CCIP solana programs

[DPA-1525](https://smartcontract-it.atlassian.net/browse/DPA-1525)
